vowpalwabbit/memory_tree.cc

```diff
@@ -0,0 +1,1358 @@
+            float router_pred = train_node(b, base, *b.examples[ec_array_index], cn); 
+            uint32_t newcn = descent(b.nodes[cn], router_pred); //updated nr or nl
+            cn = newcn; 
+        }
+    
+        if (b.oas == true)  //if useing oas as inference procedure, we just train oas here, as it's independent of the memory unit anyway'
+            train_one_against_some_at_leaf(b, base, cn, *b.examples[ec_array_index]);
+
+        if((b.nodes[cn].internal == -1) && (fake_insert == false)) //get to leaf:
+        {   
+            b.nodes[cn].examples_index.push_back(ec_array_index);
+            if (b.nodes[cn].examples_index.size() > b.max_ex_in_leaf)
+            {
+                b.max_ex_in_leaf = b.nodes[cn].examples_index.size();
+            }
+            float leaf_pred = train_node(b, base, *b.examples[ec_array_index], cn); //tain the leaf as well.
+            descent(b.nodes[cn], leaf_pred); //this is a faked descent, the purpose is only to update nl and nr of cn
+
+            //if the number of examples exceeds the max_leaf_examples, and not reach the max_nodes - 2 yet, we split:
+            if((b.nodes[cn].examples_index.size() >= b.max_leaf_examples) && (b.nodes.size() + 2 <= b.max_nodes)){
+                split_leaf(b, base, cn); 
+            }
+        }
+    }
+
+    void experience_replay(memory_tree& b, single_learner& base)
+    {
+      //return; // TODO turn this back on
+        uint32_t cn = 0; //start from root, randomly descent down! 
+	    int ec_id = random_sample_example_pop(b,cn);
+	    if (ec_id >= 0){
+            if (b.current_pass < 1)
+            	insert_example(b, base, ec_id); //unsupervised learning
+            else{
+		if (b.dream_at_update == false){
+	        	v_array<uint32_t> tmp_path = v_init<uint32_t>();
+			route_to_leaf(b, base, ec_id, 0, tmp_path, true);
+			tmp_path.delete_v();
+		}
+		else{
+                	insert_example(b, base, ec_id);
+		}
+            }
+        }
+    }
+
+    //learn: descent the example from the root while generating binary training
+    //example for each node, including the leaf, and store the example at the leaf.
+    void learn(memory_tree& b, single_learner& base, example& ec)
+    {        
+        if (b.test_mode == false){
+            b.iter++;
+            predict(b, base, ec);
+            
+            if (b.iter%5000 == 0){
+                if (b.oas == false)
+	                cout<<"at iter "<<b.iter<<", top("<<b.top_K<<") pred error: "<<b.num_mistakes*1./b.iter<<", total num queires so far: "<<b.total_num_queires<<", max depth: "<<b.max_depth<<", max exp in leaf: "<<b.max_ex_in_leaf<<endl;
+                else
+                    cout<<"at iter "<<b.iter<<", avg hamming loss: "<<b.hamming_loss*1./b.iter<<endl;
+            }
+
+            clock_t begin = clock();
+        
+            if (b.current_pass < 1){ //in the first pass, we need to store the memory:
+                example* new_ec = &calloc_or_throw<example>();
```

Comment:
Let's avoid allocation by doing this in the memory_tree& b.

Reply:
I don't understand, you wanna pre-allocate a large enough amount of memory during the initialization of memory_tree b?

Reply:
Ah, I see: is this the creation of an example to store in a leaf?

Reply:
yes,

Reply:
Carry on then :-)